nat/behavior/discovery.go

```diff
@@ -0,0 +1,366 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package behavior
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/pion/stun"
+)
+
+// Enum of DiscoverNATBehavior return values
+const (
+	NatTypeNone               = "none"
+	NatTypeFullCone           = "fullcone"
+	NatTypeRestrictedCone     = "rcone"
+	NatTypePortRestrictedCone = "prcone"
+	NatTypeSymmetric          = "symmetric"
+)
+
+// Enum of DiscoverNATMapping return values
+const (
+	MappingNone                 = "none"
+	MappingIndependent          = "independent"
+	MappingAddressDependent     = "address"
+	MappingAddressPortDependent = "addressport"
+)
+
+// Enum of DiscoverNATFiltering return values
+const (
+	FilteringIndependent = "independent"
+	FilteringAddress     = "address"
+	FilteringAddressPort = "addressport"
+)
+
+// DEFAULT_TIMEOUT for each single STUN request
+const DefaultTimeout = 3 * time.Second
+
+type stunServerConn struct {
+	conn        net.PacketConn
+	LocalAddr   net.Addr
+	RemoteAddr  *net.UDPAddr
+	OtherAddr   *net.UDPAddr
+	messageChan chan *stun.Message
+}
+
+func (c *stunServerConn) Close() error {
+	return c.conn.Close()
+}
+
+// STUN protocol compatibility errors
+var (
+	ErrResponseMessage = errors.New("error reading from response message channel")
+	ErrNoXorAddress    = errors.New("no XOR-MAPPED-ADDRESS in message")
+	ErrNoOtherAddress  = errors.New("no OTHER-ADDRESS in message")
+)
+
+// DiscoverNATBehavior returns either one of NAT_TYPE_* constants describing
+// NAT behavior in practical sense for P2P connections or error
+func DiscoverNATBehavior(ctx context.Context, address string, timeout time.Duration) (string, error) {
+	if timeout == 0 {
+		timeout = DefaultTimeout
+	}
+
+	mapping, err := DiscoverNATMapping(ctx, address, timeout)
+	if err != nil {
+		return "", err
+	}
+	switch mapping {
+	case MappingAddressDependent, MappingAddressPortDependent:
+		return NatTypeSymmetric, nil
+	case MappingNone:
+		return NatTypeNone, nil
+	}
+
+	filtering, err := DiscoverNATFiltering(ctx, address, timeout)
+	if err != nil {
+		return "", err
+	}
+	switch filtering {
+	case FilteringIndependent:
+		return NatTypeFullCone, nil
+	case FilteringAddress:
+		return NatTypeRestrictedCone, nil
+	default:
+		return NatTypePortRestrictedCone, nil
+	}
+}
+
+// DiscoverNATMapping returns either one of MAPPING_* constants describing
+// NAT mapping behavior or error
+func DiscoverNATMapping(ctx context.Context, addrStr string, timeout time.Duration) (string, error) {
```

Comment:
Again, no need to mark a string as a string.
```suggestion
func DiscoverNATMapping(ctx context.Context, address string, timeout time.Duration) (string, error) {
```

Reply:
Fixed